### PR TITLE
Analyzer revision

### PR DESCRIPTION
**Pull request by iurii-i-popov**
This PR adds revision-based cluster analyzer operations.
 [Design document](https://github.com/arangodb/documents/blob/master/DesignDocuments/03_IN_PROGRESS/AnalyzersRevisions.md)

Also this PR contains optimization for analyzers loading in case one OneShard database. If database in question is not _system than all shards reside on DBServer performing query. So Cluster-wide query is not necessary. DBServer could determine Leader shard for _analyzers collection and do local query to this only shard.

Backlog links:
https://github.com/arangodb/backlog/issues/669
https://github.com/arangodb/backlog/issues/695

### CODE REVIEW

File: arangod/Cluster/ClusterTypes.h [hunk 1]
```diff
@@ -76,6 +76,32 @@ class RebootId {
...
+  Revision getRevision() const noexcept {
+    return _revision;
+  }
+
+  Revision getBuildingRevision() const noexcept {
+    return _buildingRevision;
+  }
+
+ private:
+  Revision _revision;
```
Comment (Dronplane) [inline arangod/Cluster/ClusterTypes.h:101]:
```suggestion
  Revision _revision{0};
```

File: arangod/Cluster/ClusterTypes.h [hunk 2]
```diff
@@ -76,6 +76,32 @@ class RebootId {
...
+  Revision _buildingRevision;
```
Comment (Dronplane) [inline arangod/Cluster/ClusterTypes.h:102]:
```suggestion
  Revision _buildingRevision{0};
```

File: arangod/Cluster/ClusterInfo.cpp [hunk 3]
```diff
@@ -783,6 +785,73 @@ void ClusterInfo::loadPlan() {
...
+  //  },...
+  // }}
+
+  // Now the same for analyzers:
+  auto planAnalyzersSlice = planSlice.get("Analyzers");  // format above
+
+  if (planAnalyzersSlice.isObject()) {
+    swapAnalyzers = true;  // mark for swap even if no databases present to ensure dangling datasources are removed
+
+    for (auto const databasePairSlice : velocypack::ObjectIterator(planAnalyzersSlice)) {
```
Comment (Dronplane) [inline arangod/Cluster/ClusterInfo.cpp:801]: "databasePairSlice" is not a very good name as there could will be 2 more memebers - coordinator and rebootid. Maybe we should name it smth like  "databaseDataSlice"

File: arangod/Cluster/ClusterInfo.h [hunk 4]
```diff
@@ -287,6 +287,54 @@ class CollectionInfoCurrent {
...
+  }
+
+  static int32_t getPendingCount() noexcept {
+    return _pendingAnalyzerOperationsCount.load(std::memory_order::memory_order_relaxed);
+  }
+
+  Result start();
+  Result commit();
+  Result abort();
+private:
```
Comment (gnusi) [inline arangod/Cluster/ClusterInfo.h:312]: add whitespace

File: arangod/Cluster/ClusterInfo.h [hunk 5]
```diff
@@ -287,6 +287,54 @@ class CollectionInfoCurrent {
                              // underpins the data presented in this object
 };
 
+
+class AnalyzerModificationTransaction {
+public:
```
Comment (gnusi) [inline arangod/Cluster/ClusterInfo.h:292]: add whitespace

File: arangod/Cluster/ClusterInfo.h [hunk 6]
```diff
@@ -287,6 +287,54 @@ class CollectionInfoCurrent {
...
+  AnalyzerModificationTransaction(DatabaseID const& database, ClusterInfo* ci, bool cleanup)
+    : _clusterInfo(ci), _database(database), _cleanupTransaction(cleanup) {
+    TRI_ASSERT(_clusterInfo);
+  }
+
+  AnalyzerModificationTransaction(AnalyzerModificationTransaction const&) = delete;
+  AnalyzerModificationTransaction& operator=(AnalyzerModificationTransaction const&) = delete;
+
+  ~AnalyzerModificationTransaction() {
+    abort();
```
Comment (gnusi) [inline arangod/Cluster/ClusterInfo.h:302]: we have to ensure `abort()` is marked `noexcept` or suppress exceptions in destructor

File: arangod/Cluster/ClusterInfo.cpp [hunk 7]
```diff
@@ -5001,6 +5368,76 @@ CollectionWatcher::~CollectionWatcher() {
...
+    TRI_ASSERT(_pendingAnalyzerOperationsCount.load() == -1);
+    _pendingAnalyzerOperationsCount = 0;
+  } else {
+    TRI_ASSERT(_pendingAnalyzerOperationsCount.load() > 0);
+    _pendingAnalyzerOperationsCount.fetch_sub(1);
+  }
+  _rollbackCounter = false;
+}
+
+std::atomic<int32_t>  arangodb::AnalyzerModificationTransaction::_pendingAnalyzerOperationsCount;
```
Comment (gnusi) [inline arangod/Cluster/ClusterInfo.cpp:5450]: default `std::atomic` constructor doesn't initialize value

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 8]
```diff
@@ -504,7 +502,10 @@ arangodb::aql::AqlValue aqlFnTokens(arangodb::aql::ExpressionContext* /*expressi
                           ? server.getFeature<arangodb::SystemDatabaseFeature>().use()
                           : nullptr;
     if (sysVocbase) {
-      pool = analyzers.get(name, trx->vocbase(), *sysVocbase);
+      pool = analyzers.get(name, trx->vocbase(), *sysVocbase,
+                           expressionContext ?
+                             expressionContext->analyzersRevision() :
+                             arangodb::AnalyzersRevision::LATEST); // no context = no query => LATEST is ok
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:508]: we have to get analyzer revision from the transaction

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 9]
```diff
@@ -1333,18 +1419,16 @@ Result IResearchAnalyzerFeature::emplaceAnalyzer( // emplace
   }
 
   static const auto generator = []( // key + value generator
-    irs::hashed_string_ref const& key, // source key
-    AnalyzerPool::ptr const& value // source value
-  )->irs::hashed_string_ref {
+    irs::hashed_string_ref const& key,
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:1422]: 4 white spaces for args

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 10]
```diff
@@ -2108,6 +2300,31 @@ Result IResearchAnalyzerFeature::loadAnalyzers(
   return name; // name prefixed with vocbase (or NIL)
 }
 
+AnalyzersRevision::Ptr IResearchAnalyzerFeature::getAnalyersRevision(const irs::string_ref& vocbaseName,
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:2306]: typo `getAnalyzersRevision`

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 11]
```diff
@@ -2108,6 +2300,31 @@ Result IResearchAnalyzerFeature::loadAnalyzers(
...
+  TRI_vocbase_t* vocbase{ nullptr };
+  auto& dbFeature = server().getFeature<DatabaseFeature>();
+  vocbase = dbFeature.useDatabase(vocbaseName.empty() ? irs::string_ref(arangodb::StaticStrings::SystemDatabase) : vocbaseName);
+  if (vocbase) {
+    return getAnalyersRevision(*vocbase, forceLoadPlan);
+  }
+  return AnalyzersRevision::getEmptyRevision();
+}
+
+AnalyzersRevision::Ptr IResearchAnalyzerFeature::getAnalyersRevision(const TRI_vocbase_t& vocbase,
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:2317]: typo `getAnalyzersRevision`

File: arangod/Cluster/ClusterTypes.h [hunk 12]
```diff
@@ -76,6 +77,56 @@ class RebootId {
...
+    return _serverID;
+  }
+
+  RebootId const& getRebootID() const noexcept {
+    return _rebootID;
+  }
+
+  arangodb::velocypack::Builder toVelocyPack() const;
+
+  static Ptr getEmptyRevision() {
```
Comment (gnusi) [inline arangod/Cluster/ClusterTypes.h:117]: maybe we should think on how to protect returned object from modification

File: arangod/IResearch/IResearchAnalyzerFeature.cpp [hunk 13]
```diff
@@ -2133,6 +2426,18 @@ Result IResearchAnalyzerFeature::remove(
     WriteMutex mutex(_mutex);
     SCOPED_LOCK(mutex);
 
+    // FIXME: really strange  we don`t have this load in remove
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.cpp:2432]: right, we need to revisit this part

File: arangod/IResearch/IResearchAnalyzerFeature.h [hunk 14]
```diff
@@ -296,8 +307,25 @@ class IResearchAnalyzerFeature final
   ///////////////////////////////////////////////////////////////////////////////
   void invalidate(const TRI_vocbase_t& vocbase);
 
+  ///////////////////////////////////////////////////////////////////////////////
+  /// @brief return current known analyzers revision
+  /// @param vocbase to get revision for
+  /// @param forceLoadPlan force request to get latest available revision
+  /// @return revision number. always 0 for single server and before plan is loaded
+  ///////////////////////////////////////////////////////////////////////////////
+  AnalyzersRevision::Ptr getAnalyersRevision(const TRI_vocbase_t& vocbase, bool forceLoadPlan = false) const;
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.h:316]: typo `getAnalyzersRevision`

File: arangod/IResearch/IResearchAnalyzerFeature.h [hunk 15]
```diff
@@ -296,8 +307,25 @@ class IResearchAnalyzerFeature final
...
+
+  ///////////////////////////////////////////////////////////////////////////////
+/// @brief return current known analyzers revision
+/// @param vocbase name to get revision for
+/// @param forceLoadPlan force request to get latest available revision
+/// @return revision number. always 0 for single server and before plan is loaded
+///////////////////////////////////////////////////////////////////////////////
+  AnalyzersRevision::Ptr getAnalyersRevision(const irs::string_ref& vocbaseName, bool forceLoadPlan = false) const;
```
Comment (gnusi) [inline arangod/IResearch/IResearchAnalyzerFeature.h:324]: typo `getAnalyzersRevision`

File: arangod/StorageEngine/TransactionState.cpp [hunk 16]
```diff
@@ -54,7 +56,16 @@ TransactionState::TransactionState(TRI_vocbase_t& vocbase,
       _hints(),
       _options(options),
       _serverRole(ServerState::instance()->getRole()),
-      _registeredTransaction(false) {}
+      _registeredTransaction(false) {
+  if (isCoordinator()) {
```
Comment (gnusi) [inline arangod/StorageEngine/TransactionState.cpp:60]: we have to move this to `ClusterTransactionState::beginTransaction()`, of course we don't need `isCoordinator()` check

File: arangod/StorageEngine/TransactionState.cpp [hunk 17]
```diff
@@ -54,7 +56,16 @@ TransactionState::TransactionState(TRI_vocbase_t& vocbase,
...
+    // for cluser operation we need to set analyzersRevision
+    if (_vocbase.server().hasFeature<arangodb::iresearch::IResearchAnalyzerFeature>()) {
+      _analyzersRevision = _vocbase.server()
+        .getFeature< arangodb::iresearch::IResearchAnalyzerFeature>()
+        .getAnalyzersRevision(_vocbase, true)->getRevision();
+    }
+  }
```
Comment (gnusi) [inline arangod/StorageEngine/TransactionState.cpp:67]: instead of lookup feature twice, I'd better do something like:
```
  arangodb::iresearch::IResearchAnalyzerFeature* feature = nullptr;

  try {
    feature = _vocbase.server().getFeature< arangodb::iresearch::IResearchAnalyzerFeature>();
  } catch (...) {
    return;
  }

  _analyzersRevision = feature->getAnalyzersRevision(_vocbase, true)->getRevision();
```

File: arangod/StorageEngine/TransactionState.cpp (hunk 16 above)
Comment (Dronplane) [reply arangod/StorageEngine/TransactionState.cpp:60]: Good point about ClusterTransactionState. But beginTransaction is to late. Still need to do it in ctor, as optimization step need to know revision.

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/10011/

File: arangod/Cluster/ClusterTypes.h [hunk 18]
```diff
@@ -76,6 +79,62 @@ class RebootId {
...
+
+struct AnalyzersRevision {
+ public:
+  using Revision = uint64_t;
+  using Ptr = std::shared_ptr<AnalyzersRevision const>;
+
+  static constexpr Revision LATEST = std::numeric_limits<uint64_t>::max();
+  static constexpr Revision MIN = 0;
+
+  struct Key { // passkey idiom
```
Comment (gnusi) [inline arangod/Cluster/ClusterTypes.h:95]: can't we just make constructor private?

File: arangod/StorageEngine/TransactionState.cpp [hunk 19]
```diff
@@ -291,6 +292,14 @@ void TransactionState::setExclusiveAccessType() {
   _type = AccessMode::Type::EXCLUSIVE;
 }
 
+void TransactionState::acceptAnalyzersRevision(AnalyzersRevision::Revision analyzersRevision) noexcept {
+  // only init from default allowed! Or we have problem -> different analyzersRevision in one transaction
+  LOG_TOPIC_IF("9127a", FATAL, Logger::AQL, (_analyzersRevision != analyzersRevision && _analyzersRevision != AnalyzersRevision::MIN))
+    << " Changing analyzers revision for transaction from " << _analyzersRevision << " to " << analyzersRevision;
+  TRI_ASSERT(_analyzersRevision == analyzersRevision || _analyzersRevision == AnalyzersRevision::MIN);
```
Comment (gnusi) [inline arangod/StorageEngine/TransactionState.cpp:299]: doesn't FATAL terminate process?

Comment (Dronplane) [reply arangod/StorageEngine/TransactionState.cpp:299]: No, It just forces logger flush

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/10030/

File: arangod/VocBase/VocbaseInfo.cpp [hunk 20]
```diff
@@ -149,9 +148,10 @@ void CreateDatabaseInfo::toVelocyPack(VPackBuilder& builder, bool withUsers) con
   builder.add(StaticStrings::DatabaseName, VPackValue(_name));
   builder.add(StaticStrings::DataSourceSystem, VPackValue(_name == StaticStrings::SystemDatabase));
 
-  if (ServerState::instance()->isCoordinator()) {
+  if (ServerState::instance()->isCoordinator() ||
+      ServerState::instance()->isDBServer()) {
```
Comment (kvahed) [inline arangod/VocBase/VocbaseInfo.cpp:152]: Why is this necessary again?

Comment (Dronplane) [reply arangod/VocBase/VocbaseInfo.cpp:152]: That is for  OneShard optimization. Without this DBServer can`t determine if vocbase is OneShard and apply optimization for analyzers loading

**Comment by Dronplane** [comment]
https://jenkins.arangodb.biz/view/PR/job/arangodb-matrix-pr/10043/
